source/common/stats/symbol_table_impl.cc

```diff
@@ -126,69 +126,109 @@ SymbolVec SymbolTableImpl::Encoding::decodeSymbols(const SymbolTable::Storage ar
       // var-length encode the size of the literal string prior to the data.
-      ASSERT(size > 1);
-      ++array;
-      --size;
-      std::pair<uint64_t, size_t> length_consumed = decodeNumber(array);
+      ASSERT(size_ > 1);
+      ++array_;
+      --size_;
+      std::pair<uint64_t, size_t> length_consumed = decodeNumber(array_);
       uint64_t length = length_consumed.first;
-      array += length_consumed.second;
-      size -= length_consumed.second;
-      ASSERT(size >= length);
-      string_view_token_fn(absl::string_view(reinterpret_cast<const char*>(array), length));
-      size -= length;
-      array += length;
-    } else {
-      std::pair<uint64_t, size_t> symbol_consumed = decodeNumber(array);
-      symbol_token_fn(symbol_consumed.first);
-      size -= symbol_consumed.second;
-      array += symbol_consumed.second;
+      array_ += length_consumed.second;
+      size_ -= length_consumed.second;
+      ASSERT(size_ >= length);
+      string_view_ = absl::string_view(reinterpret_cast<const char*>(array_), length);
+      size_ -= length;
+      array_ += length;
+      return Type::StringView;
     }
+    std::pair<uint64_t, size_t> symbol_consumed = decodeNumber(array_);
+    symbol_ = symbol_consumed.first;
+    size_ -= symbol_consumed.second;
+    array_ += symbol_consumed.second;
+    return Type::Symbol;
   }
-}
 
-bool StatName::startsWith(StatName symbolic_prefix) const {
-  bool ret = true;
-  std::vector<Symbol> prefix_symbols;
+  /** @return the current string_view -- only valid to call if next()==Type::StringView */
+  absl::string_view stringView() const { return string_view_; }
+
+  /** @return the current symbol -- only valid to call if next()==Type::Symbol */
+  Symbol symbol() const { return symbol_; }
 
-  // Decode the prefix as a StatNameVec.
-  SymbolTableImpl::Encoding::decodeTokens(
-      symbolic_prefix.data(), symbolic_prefix.dataSize(),
-      [&prefix_symbols](Symbol symbol) { prefix_symbols.push_back(symbol); },
-      [&ret](absl::string_view) { ret = false; });
+private:
+  const uint8_t* array_;
+  size_t size_;
+  absl::string_view string_view_;
+  Symbol symbol_;
+};
 
-  // If there any dynamic components, our string_view lambda will be called, and
-  // then we'll return false for simplicity. We don't have a current need for
-  // prefixes to be expressed dynamically, and handling that case would add
-  // complexity.
-  if (!ret) {
-    return false;
+void SymbolTableImpl::Encoding::decodeTokens(
+    const SymbolTable::Storage array, size_t size,
+    const std::function<void(Symbol)>& symbol_token_fn,
+    const std::function<void(absl::string_view)>& string_view_token_fn) {
+  TokenIter iter(array, size);
+  TokenIter::Type type;
+  while ((type = iter.next()) != TokenIter::Type::End) {
+    if (type == TokenIter::Type::StringView) {
+      string_view_token_fn(iter.stringView());
+    } else {
+      symbol_token_fn(iter.symbol());
```

Comment:
Assert(type == TokenIter::Type::Symbol) to guard against a new type being added in some distant future?